presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1226,7 +1227,7 @@ private PlanWithProperties arbitraryDistributeUnion(
                 List<List<VariableReferenceExpression>> distributedOutputLayouts)
         {
             // TODO: can we insert LOCAL exchange for one child SOURCE distributed and another HASH distributed?
-            if (getNumberOfTableScans(distributedChildren) == 0) {
+            if (getNumberOfTableScans(distributedChildren) == 0 && isPartitioningCompatible(extractRemoteExchangePartitioningHandles(distributedChildren))) {
```

Comment:
is this fix related to materialized exchanges or could it have been an issue before?

Reply:
It could've been an issue even before. It happens when there's a JOIN over union of incompatible partitioning distribution. For example when doing join over union of aggregations over bucketed tables.

Reply:
@arhimondr : Curious why it's not reported before? :)

Reply:
Replicated joins are rare and bucketed tables are rare. And the query shape also not that common.